enhancements/storage/csi-driver-operator-merge.md

```diff
@@ -0,0 +1,530 @@
+  ├── pkg
+  │   ├── aws-ebs
+  │   │   └── # AWS EBS specific-code
+  │   └── common
+  │       └── starter.go # and any other code shared by all CSI drivers
+  ├── test
+  │   └── e2e
+  │       └── aws-ebs
+  │           └── manifest.yaml
+  ├── vendor
+  ├── Dockerfile.aws-ebs
+  ├── Dockerfile.aws-ebs-test
+  ├── go.mod
+  └── go.sum
+  ```
+
+There will be two major points when things can break:
+
+1. When we switch building of `ose-aws-ebs-csi-driver-operator-container` from openshift/aws-ebs-csi-driver-operator to
+   openshift/csi-operator. Since the code will be 100% the same, we think it's safe to do. Any revert must be
+   coordinated with ART in ocp-build-data and openshift/release repositories.
+2. After our refactoring, we will switch building of `Dockerfile.aws-ebs` from `legacy/aws-ebs-csi-driver-operator/cmd/`
+   to `cmd/`. At this time we will have CI that should catch any breakage. Any revert must be done in
+   openshift/csi-operator repo, which is under our control.
+
+All CSI driver operators will be merged in a similar way, i.e. in `legacy/` directory first.
+
+Drawbacks:
+
+* When anything goes wrong in switching the source repositories, it's ART who will need to roll back things, we cannot
+  do it ourselves.
+
+Advantages:
+
+* Keeping image names, i.e. no new Comet repos (distgits, brew packages, etc.)
+* Harder (but not impossible) to test before the switch. In ideal case, we're switching just the source repository
+  of the same image, the actual operator code should be the same.
+
+#### Gradual merge
+
+We will merge the operators one by one, starting with AWS EBS. Second will be Azure Disk and Azure File, as we need to
+support HyperShift for them and we want to share code with AWS EBS. The rest will follow as time allows, possibly over
+multiple OCP releases.
+
+Very provisional and optimistic plan:
+
+* AWS EBS, Azure Disk and Azure File in 4.15.
+* The rest in 4.16 or later.
+
+#### Generated YAML files
+
+All CSI driver YAML files look very similar today. We plan to generate them from a single set of templates, so that we
+don't need to maintain them separately. Exact details about the generator are TBD. We want something like kustomize,
+but better integrated with the operator code.
+
+### Risks and Mitigations
+
+* Since each CSI driver operator is different, it may be hard to merge all their code. Some differences are:
+  * Shared Resource (and maybe others) deploy an extra validating webhook.
+  * Cinder (and maybe others) syncs Secrets to a different format.
+  * Manila installs the CSI driver only optionally.
+  * AWS sync CA-bundle from a different namespace.
+  * vSphere interacts with vSphere API a lot.
+  * AWS EFS and GCP Filestore are OLM operators.
+  * And many others.
+
+  All the operators use shared code from library-go already, so merging them to a single repository will not make them
+  worse. Still, it may be more difficult to share even more code e.g. for HyperShift.
```

Comment:
Do you think there are conflicting `go.mod` among the operators? How would you handle it if that was the case?

Reply:
Just saw you mention it in the `Drawbacks` section below. Still interested in knowing how we would handle conflicts if we had some. Do you see us forking libraries for example?

Reply:
No, no forking if possible. In the absolutely worst case we may end up with multiple vendor directories / go.mods. But, it should be possible to have all cloud SDKs in a single vendor - they were there once Kubernetes had all cloud providers in-tree. I remember there were issues around that, but they solved them somehow.